[CIAPP-2205] Add Git metadata troubleshooting section (#11038)
DataDog/documentation: content/en/continuous_integration/troubleshooting.md

kind: documentation
---

### Your Jenkins instance is instrumented, but Datadog isn't showing any data

1. Make sure that at least one pipeline has finished executing. Pipeline execution information is only sent after the pipeline has finished.
2. Make sure the Datadog Agent host is properly configured and it's reachable by the Datadog Plugin. You can test connectivity by clicking on the **Check connectivity with the Datadog Agent** button on the Jenkins plugin configuration UI.

> 3. If you still don't see any results, [contact Support][3] and we'll help you troubleshoot.

Reviewer: ```suggestion
3. If you still don't see any results, [contact Support][3] for troubleshooting help.
```

3. For languages other than Swift, make sure the Datadog Agent is running on the host where tests are run (accessible at `localhost:8126`), or if accessible on another hostname or port, make sure you run your tests with the appropriate Agent hostname set in the `DD_AGENT_HOST` and the appropriate port in `DD_TRACE_AGENT_PORT` environment variables. You can activate [debug mode][8] in the tracer to check if it's able to connect to the Agent.
4. If you still don't see any results, [contact Support][3] for troubleshooting help.

### Data appears in Test Runs but not Tests

If you can see test results data in the **Test Runs** tab, but not the **Tests** tab, Git metadata (repository, commit and/or branch) is probably missing. To confirm this is the case, open a test execution in the [Test Runs][4] section, and check that there is no `git.repository_url`, `git.commit.sha`, or `git.branch`. If these tags are not populated, nothing will show on the [Tests][5] section.

> 1. Tracers will first try to fetch Git metadata using the local `.git` folder by executing `git` commands. This is the preferred approach, as it will populate all Git metadata fields, including commit message, author and committer information.

Reviewer: ```suggestion
1. Tracers will first try to fetch Git metadata using the local `.git` folder by executing `git` commands. This is the preferred approach, as it will populate all Git metadata fields, including commit message, author and committer information. Ensure the `.git` folder is present and the `git` binary is installed and in `$PATH`.
```

> 2. If the `.git` folder is not present, or the `git` binary is not installed, tracers will fallback to use the environment variables set by the CI provider (as listed in the [Running tests inside a container][6] section) to collect Git information. This will populate, at least, repository, commit hash and branch information.

Reviewer: ```suggestion
2. If the `.git` folder is not present, or the `git` binary is not installed, tracers will fallback to use the environment variables set by the CI provider to collect Git information. See the [Running tests inside a container][6] page for a list of environment variables to set with the provider. This will populate, at least, repository, commit hash and branch information.
```

Author: The env vars are not to be set by the user, just read by the tracer. What this (or similar):
```suggestion
2. If the `.git` folder is not present, or the `git` binary is not installed, tracers will fallback to use the environment variables set by the CI provider to collect Git information. See the [Running tests inside a container][6] page for a list of environment variables that the tracer will attempt to read for each supported CI provider. This will populate, at least, repository, commit hash and branch information.
```

kind: documentation
---

> ### I instrumented my Jenkins instance but I don't see any data in Datadog

Reviewer 2: There is a troubleshooting section in Jenkins instructions. Would that section better for this?

If not, this is Jenkins specific and the other ones are more generic. It could be misleading because if the user doesn't read carefully, it might think that this applies to everything. Maybe different sections such as: `General`,  `Pipelines - Jenkins`, or something similar.

Author: This particular "issue" is not present on the Jenkins section. However, I would try to consolidate all troubleshooting here going forward, and if it gets too messy, we can add different sections.

Reviewer 2: In that case, maybe we should consider moving troubleshooting from Jenkins to here.

3. If you still don't see any results, [contact Support][3] and we'll help you troubleshoot.

### I instrumented my tests but I don't see any data in Datadog

1. Check the _Compatibility_ section of the page of the language you are instrumenting, and make sure the testing framework you are using is supported.
2. Check if you see any test results in the [Test Runs][4] section. If you do see results there, but not in the [Tests][5] section, Git information  might be missing. If this is this case, read the next section to troubleshoot it.
3. For all languages (except [Swift][7]), make sure the Datadog Agent is running in the host where tests are run (accessible at `localhost:8126`), or if accessible on another hostname and/or port, make sure you run your tests with the appropriate agent hostname set in the `DD_AGENT_HOST` and the appropriate port in `DD_TRACE_AGENT_PORT` environment variables. You can activate [debug mode][8] in the tracer to check if it's being able to connect to the agent.
4. If you still don't see any results, [contact Support][3] and we'll help you troubleshoot.

### I see my tests in the "Test Runs" tab, but not in the "Tests" tab

> This is most probably due to missing Git metadata (repository, commit and/or branch). To confirm this is the case, open a test execution on the [Test Runs][4] section, and check that there is no `git.repository_url`, `git.commit.sha` or `git.branch`. If these tags are not populated, nothing will show on the [Tests][5] section.

Reviewer 2: 💯

1. Tracers will first try to fetch Git metadata using the local `.git` folder by executing `git` commands. This is the preferred approach, as it will populate all Git metadata fields, including commit message, author and committer information.
2. If the `.git` folder is not present, or the `git` binary is not installed, tracers will fallback to use the environment variables set by the CI provider (as listed in the [Running tests inside a container][6] section) to collect Git information. This will populate, at least, repository, commit hash and branch information.

> 3. If no CI provider environment variables are found, tests results will be sent with no Git metadata.

Reviewer: This describes what the software does, but not what the reader should do to troubleshoot. What should they check or fix? If they find themselves in step 3, does that mean they just have to live with no data on the Tests tab?

Author: We are currently working on a way for the user to be able to manually define Git data if (1) and (2) are impossible to do on their setup, which would replace this last step.

The reason to explain how the tracer works is for them to be able to make changes on their setup to allow the tracer to automatically detect Git data. We could rephrase this in a more "troubleshooting" way, so instead of saying what the tracer will do, we can say what they need to do (for example, "make sure the `.git` folder is present and the `git` binary is installed and in `$PATH` or something like that)

Reviewer: I agree that telling them how the tracer works is valuable here. How about these rewordings to make it clearer what they can do? Let me know what you think.

Author: Just left a small comment on the second suggestion, otherwise LGTM

Reviewer: Great! I committed my first suggestion and your second suggestion, and will 🚢

> ### I have another issue not described here

Reviewer: ```suggestion
### Need further help?
```

> If that's the case, please [contact Support][3] and we'll help you troubleshoot your issue.

Reviewer: ```suggestion
If you have another issue, or the above solutions don't work, [contact Support][3] for troubleshooting help.
```

[1]: https://app.datadoghq.com/ci/pipeline-executions
[2]: https://app.datadoghq.com/ci/pipelines
[3]: /help/
[4]: https://app.datadoghq.com/ci/test-runs
[5]: https://app.datadoghq.com/ci/test-services
[6]: /continuous_integration/setup_tests/containers/

> [7]: /continuous_integration/setup_tests/swift/

Reviewer: ```suggestion
```
